beacon-chain/sync/block_batcher.go

```diff
@@ -0,0 +1,210 @@
+	nb.seq, nb.nonseq, nb.err = bb.cf.filter(ctx, rob)
+
+	// Decrease allowed blocks capacity by the number of streamed blocks.
+	bb.limiter.add(stream, int64(1+nb.end.SubSlot(nb.start)))
+	bb.current = &nb
+	return *bb.current, true
+}
+
+func (bb *blockRangeBatcher) genesisBlock(ctx context.Context) (blocks.ROBlock, error) {
+	b, err := bb.db.GenesisBlock(ctx)
+	if err != nil {
+		return blocks.ROBlock{}, err
+	}
+	htr, err := b.Block().HashTreeRoot()
+	if err != nil {
+		return blocks.ROBlock{}, err
+	}
+	return blocks.NewROBlockWithRoot(b, htr)
+}
+
+type blockBatch struct {
+	start  primitives.Slot
+	end    primitives.Slot
+	seq    []blocks.ROBlock
+	nonseq []blocks.ROBlock
+	err    error
+}
+
+func newBlockBatch(start, reqEnd primitives.Slot, size uint64) (blockBatch, bool) {
+	if start > reqEnd {
+		return blockBatch{}, false
+	}
+	nb := blockBatch{start: start, end: start.Add(size - 1)}
+	if nb.end > reqEnd {
+		nb.end = reqEnd
+	}
+	return nb, true
+}
+
+func (bat blockBatch) next(reqEnd primitives.Slot, size uint64) (blockBatch, bool) {
+	if bat.sequenceBroken() {
+		return blockBatch{}, false
+	}
+	return newBlockBatch(bat.end.Add(1), reqEnd, size)
+}
+
+// sequence returns the list of sequential, canonical blocks read from the db.
+func (bb blockBatch) sequence() []blocks.ROBlock {
+	return bb.seq
+}
+
+// sequenceBroken is used to determine if there was a break in the chain of canonical blocks as read from the db.
+// If true, code using the blockBatch should stop serving additional batches of blocks.
+func (bb blockBatch) sequenceBroken() bool {
+	return len(bb.nonseq) > 0
+}
+
+func (bb blockBatch) error() error {
+	return bb.err
+}
+
+type canonicalChecker func(context.Context, [32]byte) (bool, error)
+
+type canonicalFilter struct {
+	prevRoot  [32]byte
+	canonical canonicalChecker
+}
+
+// filters all the provided blocks to ensure they are canonical and strictly linear.
+func (cf *canonicalFilter) filter(ctx context.Context, blks []blocks.ROBlock) ([]blocks.ROBlock, []blocks.ROBlock, error) {
+	blks = sortedUniqueBlocks(blks)
+	seq := make([]blocks.ROBlock, 0, len(blks))
+	nseq := make([]blocks.ROBlock, 0)
+	for i, b := range blks {
+		cb, err := cf.canonical(ctx, b.Root())
+		if err != nil {
+			return nil, nil, err
+		}
+		if !cb {
+			continue
+		}
+		// prevRoot will be the zero value until we find the first canonical block in the stream seen by an instance
+		// of canonicalFilter. filter is called in batches; prevRoot can be the last root from the previous batch.
+		first := cf.prevRoot == [32]byte{}
+		// We assume blocks are processed in order, so the previous canonical root should be the parent of the next.
+		if !first && cf.prevRoot != b.Block().ParentRoot() {
+			// If the current block isn't descended from the last, something is wrong. Append everything remaining
+			// to the list of non-sequential blocks, and stop building the canonical list.
+			nseq = append(nseq, blks[i:]...)
+			break
+		}
```

Comment:
How could this happen? If every block is canonical, then this should never happen right?

If I understand it correctly, every canonical block would build upon the previous, highest slot, canonical block.

Reply:
My terminology is confusing here. I mean sequential in terms of ancestry, not slot number.

Overall this method is mega paranoid, because if there are gaps between canonical blocks, that means there is something wrong with the db or forkchoice. I'm not sure if I would have decided to write such a deep check if I was building this from scratch, but this preserves behavior of the existing code I refactored.

Reply:
I updated some variable names, let me know if you've got additional thoughts on how to make it more clear.